corehq/motech/fhir/tests/test_fhir_views.py

```diff
@@ -208,32 +193,71 @@ def test_search(self):
-        ):
-            request = _get_request(url, self.django_user, self.api_key.key)
-            is_authenticated = LoginAuthentication().is_authenticated(request)
-            self.assertTrue(is_authenticated)
+class APIAuthenticationTests(BaseFHIRViewTest):
+    """
+    Tests to confirm the authentication requirements of the FHIR API.
+    """
 
-    def test_superuser_not_authenticated(self):
-        self.django_user.is_superuser = True
-        self.django_user.save()
+    api_endpoints = (
+        reverse(
+            "fhir_get_view",
+            args=[DOMAIN, FHIR_VERSION, "Patient", PERSON_CASE_ID],
+        ),
+        reverse(
+            "fhir_search",
+            args=[DOMAIN, FHIR_VERSION, "Observation"],
+        ),
+    )
 
-        for url in (
-            reverse("fhir_get_view",
-                    args=[DOMAIN, FHIR_VERSION, "Patient", PERSON_CASE_ID]),
-            reverse("fhir_search",
-                    args=[DOMAIN, FHIR_VERSION, "Observation"])
-        ):
-            request = _get_request(url, self.django_user)
-            is_authenticated = LoginAuthentication().is_authenticated(request)
-            self.assertFalse(is_authenticated)
+    @flag_enabled('FHIR_INTEGRATION')
+    def test_api_key_auth(self):
+        self.client.login(username=USERNAME, password=PASSWORD)
+        for url in self.api_endpoints:
+            response = self.client.get(
+                url,
+                HTTP_AUTHORIZATION=f'ApiKey {USERNAME}:{self.api_key.key}',
+            )
+            self.assertIsInstance(response, JsonResponse)
+
+    def test_api_key_auth_without_fhir_integration(self):
+        self.client.login(username=USERNAME, password=PASSWORD)
+        for url in self.api_endpoints:
+            response = self.client.get(
+                url,
+                HTTP_AUTHORIZATION=f'ApiKey {USERNAME}:{self.api_key.key}',
+            )
+            self.assertIsInstance(response, HttpResponseNotFound)
+            self.assertHQStandard404Page(response)
+
+    def test_basic_auth(self):
+        self.client.login(username=USERNAME, password=PASSWORD)
+        b64_bytes = b64encode(f'{USERNAME}:{PASSWORD}'.encode('utf8'))
+        basic_auth = b64_bytes.decode('utf8')
+        for url in self.api_endpoints:
+            response = self.client.get(
+                url,
+                HTTP_AUTHORIZATION=f'Basic {basic_auth}',
+            )
+            self.assertIsInstance(response, HttpUnauthorized)
+
+    def test_superuser_login(self):
+        with user_is_superuser(self.django_user):
+            self.client.login(username=USERNAME, password=PASSWORD)
+            for url in self.api_endpoints:
+                response = self.client.get(url)
+                self.assertIsInstance(response, HttpUnauthorized)
+
+
+@contextmanager
+def user_is_superuser(user):
+    is_superuser = user.is_superuser
+    try:
+        user.is_superuser = True
```

Comment:
Should we avoid doing any saves if user is already a superuser?
But is this is even used anywhere apart from `test_superuser_login` that its worth adding as a context manager?

Reply:
> Should we avoid doing any saves if user is already a superuser?

Good call. In fact, we don't ever need to save. e819651b48a785587e8b942bfafc887b969b3277

> ... that its worth adding as a context manager?

**Trigger warning:** Contentious minority opinions about Python ahead.

This touches on a topic I have "mild" feelings about: JUnit. JUnit is perfect for Java. But it's not for Python, and the `unittest` standard library is, imho, a mistake that we have dealt with by using cludges like `nose` and `pytest` to make it more bearable. My minority opinion: Python should have used a framework like [Attest](https://pythonhosted.org/Attest/), written by the [authors](http://www.pocoo.org/) of Sphinx, Flask, Jinja, and Click. Like Flask, it is small, Pythonic, and beautiful, but it is effectively dead. (Its last release was in 2011.)

JUnit uses `setUp` and `tearDown` because, at the time, Java didn't have context managers. ([It does now](https://www.oracle.com/technical-resources/articles/java/trywithresources.html), modeled after Python's.) One-line functions are not uncommon in HQ, and I think that's a good thing; they make the calling code more readable.

In my perfect universe, we would use small context managers more, for more readable calling code. And we'd use them in tests all the time, instead of the horrifying [taxidermied Java-Python monster](https://www.youtube.com/watch?v=MeK40LXEIO4) that is the `TestCase` class with its ridiculous `setUpWhatever()` and `assertThisSpecificConditionBecauseJava()` methods that we're stuck with.

`</rant>`

`<summary>`
I love tiny context managers. They're like adorable baby animals, except they clean up after themselves.
`</summary>`

Reply:
Fun fact: Python 3.10 is bringing an [improved `with` statement](https://bugs.python.org/issue12782) that supports chained, transactional contexts:

```python
with (
    context_manager("foo") as context,
    another_context_manager(context),
):
    ...
```

Reply:
Thanks for sharing that Norman.

> In my perfect universe, we would use small context managers more, for more readable calling code

Ya totally not against context managers. They are great and useful. I was thinking that since its used in a single place here and is just a couple of lines, you could just set and unset is_superuser in the test itself, plus since we don't need to save the user at all.

(this is non blocking)